[WIP] Feature integration sendinblue

RignonNoel opened the pull request:
| Q                                                      | R
| ------------------------------------------ | -------------------------------------------
| Type of contribution ?                      | Feature
| Tickets (_issues_) concerned               | fixed #84

---

I've begin an integration of `SendinBlue` with the new API v3, i took the choice to directly call the API and to not use the official python client implementation as recommended by `django-anymail` documentation.

Please feel free to review my code, contribute or just help me with some tips or good idea.

I will continue this work soon 👍 
 
##### What have you changed ?
-  [x] Link the issue to this PR
-  [x] Add my name in `AUTHORS.txt`
-  [x] Create the Backend and Payload classes
-  [x] Add or modify the attached tests
-  [x] Add or modify the attached documentation
-  [x] Test with an account of the ESP (Thank for the great integration testing ❤️ )
-  [x] Add templated transactional email support
-  [x] Add tag support

##### How did you change it ?
I've based my work on [this awaiting PR for Mailjet](https://github.com/anymail/django-anymail/blob/da25eb03b64d28d0a4effcf4fe014cd86ac9a7fd/anymail/backends/mailjet.py) and the `Sendgrid` integration.

##### References

 1. [SendinBlue's general documentation for API v3](https://developers.sendinblue.com/docs/getting-started)
 2. [SensinBlue's dynamic test for API v3](https://developers.sendinblue.com/v3.0/reference#sendtransacemail-1)
 3. [SendinBlue's official client API v3](https://github.com/sendinblue/APIv3-python-library)

medmunds commented:
Very nice start—thanks for this!

I'll add a few specific notes inline, but overall this is looking good.

Some of their API documentation seems a little, um, under-specified, so there will likely be some experimentation required to get all the details right. (In some cases, their [v2 docs](https://apidocs.sendinblue.com/tutorial-sending-transactional-email/) might provide some direction; although the v3 API is different, I'm guessing it shares a lot of the same underlying platform capabilities and limitations.)

After you get the basic transactional send working, there are some additional Anymail send features that should probably be supported:
* `template_id` and `merge_data`: SiB's [transactional-template](https://developers.sendinblue.com/v3.0/reference#sendtemplate) sending
* `tags` and `metadata`: looks like maybe special X-Mailin-custom and X-Mailin-tag headers can be used for these (based on v2 docs)

medmunds commented on anymail/backends/sendinblue.py:
```diff
@@ -0,0 +1,144 @@
+        # SendinBlue doesn't give any detail on a success
+        # https://developers.sendinblue.com/docs/responses
+
+        return dict()
```
This *must* return a dict mapping recipient email address (addr_spec) to an AnymailRecipientStatus for each message recipient. Since the best we know here is "probably sent", you could use [code like this](https://github.com/anymail/django-anymail/blob/v1.4/anymail/backends/mailgun.py#L54-L56) to mark all recipients "queued".

The AnymailRecipientStatus.message_id is some sort of tracking ID that can be used to correlate this send with later webhook status events. It's not documented, but a quick test suggests SiB's response is actually json, and *does* include a usable tracking id: `{"messageId":"<201802182116.25729519001@smtp-relay.mailin.fr>"}`.

RignonNoel replied:
Exactly responses i've on my tests look like this one :
```
{
    'code': 'success',
    'message': 'Email sent successfully',
    'data': {
        'message-id': '<201802182329.73817096184@smtp-relay.sendinblue.com>'
    }
}
```

Thanks for the link on the mailgun example. It was one of my point on my todo-list of research 👍

medmunds commented on anymail/backends/sendinblue.py:
```diff
@@ -0,0 +1,144 @@
+        assert recipient_type in ["to", "cc", "bcc"]
+        if emails:
+            # Normally, exactly one "personalizations" entry for all recipients
+            # (Exception: with merge_data; will be burst apart later.)
```
(This comment doesn't apply to SiB, I think.)

medmunds commented on anymail/backends/sendinblue.py:
```diff
@@ -0,0 +1,144 @@
+            self.data['replyTo'] = self.email_object(emails[0])
+
+    def set_extra_headers(self, headers):
+        self.data['headers'] = [header for header in headers]
```
SiB expects an "object" here, but doesn't say much more. I'm guessing it's a mapping of header field-name to value, which is what the `headers` param already is. So you can probably just do `self.data['headers'] = headers`.

RignonNoel replied:
The documentation of v2 have more detail on the `headers` parameter :

|Input Parameter | Description|
| -- | -- |
| headers | The headers will be sent along with the mail  headers in original email. Example: array(“Content-Type”=>”text/html;  charset=iso-8859-1″). You can use commas to separate multiple headers  [Optional]. |

I guess it's just the same thing.

medmunds commented on anymail/backends/sendinblue.py:
```diff
@@ -0,0 +1,144 @@
+
+    def add_attachment(self, attachment):
+        """Converts attachments to SendinBlue API {name, base64} array"""
+
```
I can't find any info on SiB support for inline attachments/inline images. (May need to contact them for clarification.) If they truly don't support inlines, we'd need `if attachment.inline:   self.unsupported_feature("inline attachments")` here.

(Also, I'm really surprised they don't have a way to specify the attachment content-type; I'm wondering if they always just guess that from the name. If so, that becomes a "quirks and limitations" entry for the docs.)

RignonNoel replied:
On the API v2 we have this little part of the documentation.
Maybe the v3 keep it functionnal.. We will need to test i guess.

|Input Parameter | Description|
| -- | -- |
| html | Body of the message (HTML version) [Mandatory]. To send inline images, use < img src="{YourFileName.Extension}" alt="image" border="0" >, the src attribute value inside {} (curly braces) should be same as the filename used in inline_image parameter |
| inline_image | Pass your inline image/s filename & its  base64 encoded chunk data as an associative array. Example:  array(“YourFileName.Extension”=>”Base64EncodedChunkData”). You can  use commas to separate multiple inline images [Optional]. |

medmunds commented on anymail/backends/sendinblue.py:
```diff
@@ -0,0 +1,144 @@
+            'content': attachment.b64content,
+        }
+
+        self.data.setdefault("attachment", []).append(att)
```
Every payload should implement `set_esp_extra` to allow overrides and esp-specific parameters. Usually just dict.update into self.data works fine. (Anymail utils.update_deep can also be used.) [Example](https://github.com/anymail/django-anymail/blob/v1.4/anymail/backends/mailgun.py#L198-L199).

RignonNoel replied:
Done. I've found no one information for the extra header on the v3 so i've not included some check in the code.

> Note:
“X-Mailin-IP” will not be sent along with the mail headers in original email.
“X-param1″, “X-param2″, “X-Mailin-custom” and “X-Mailin-Tag” will be sent along with the mail headers in original email.
“X-Mailin-custom” and “X-Mailin-Tag” will also be sent along with the JSON response of webhooks, when some event occurs. Also, the “X-Mailin-Tag” will be received as only “tag” in the JSON response.

[API v2 reference](https://apidocs.sendinblue.com/tutorial-sending-transactional-email/)

RignonNoel commented:
Hi @medmunds and thanks a lot for all your comments! 

---

I've check the python client API for the v3 and i find [the part of the code](https://github.com/sendinblue/APIv3-python-library/blob/2c74a142582f23f265a2e3f433e5516a4a13a3cb/sib_api_v3_sdk/api/smtp_api.py#L1086-L1182) about transactional email without template. Unfortunately, no trace of the `inline_image` feature of the v2... 

I will ask the tech team of SendinBlue by email tomorrow.

---

For the `template` feature i'm not sure what's the best architecture to do it considering the fact that SendinBlue doesn't use the same url for templated email.

```
https://api.sendinblue.com/v3/smtp/templates/templateId/send
```

Do you have some documentation about Django's email backend i can read to understand how `set_template_id` is called ? I do not understand all the call stack used to make anymail's backend works. 

Otherwise it does not matter i'm just going to read django's documentation directly 😃

medmunds commented:
> For the template feature i'm not sure what's the best architecture to do it considering the fact that SendinBlue doesn't use the same url for templated email.

You can add logic to your Payload.get_api_endpoint. Examples in the [Postmark backend](https://github.com/anymail/django-anymail/blob/v1.4/anymail/backends/postmark.py#L109-L114) and  [Mandrill backend](https://github.com/anymail/django-anymail/blob/v1.4/anymail/backends/mandrill.py#L74-L78).

RignonNoel commented:
I just worked on the integration of templated email but their API seems really inconsistent on some part :

 - Fields `to`, `cc` and `bcc` are renamed `emailTo`, `emailCc` and `emailBcc` in case you use a `template_id`
 - These fields doesn't take the same kind of object inside:
   - `to` : `[{"email": "to1@example.com", "name": "My name"}]`
   - `emailTo`: `["to1@example.com"]`
 - `replyTo` keep the same name but do not have the same content also

Considering this really poor conception of their API i've made a custom function to rename fields and change the content. My custom function is called in the `serialize_data()` in case we have a `template_id`.

---

I've thought about manage two variable `self.data` and `self.data_template` and keep them both updated on EACH function but i was not sure about which part of the system have access to `self.data` since the variable doens't use the `private` practice of python (ie: `self._data`).

The actual implementation i've made required to change the new custom function in case we change the `self.data` structure (ex: new SiB's API documentation). It's not a big problem since v3 is officially release, but it's not the best way to do it.

medmunds commented:
@RignonNoel awesome! Sorry for the delay responding.

The only significant piece missing for the backend right now is backend (non-integration) tests. These mock out requests to verify that Anymail is trying to send the expected parameters to SendinBlue's API, for a variety of different cases. I usually just copy the tests for an ESP with a similar API (e.g., [test_mailjet_backend.py](https://github.com/anymail/django-anymail/blob/feature_mailjet_v3.1/tests/test_mailjet_backend.py)) and modify them. Let me know if that's enough to go on, or if I can provide any help.

Re: their template API using a totally different format for to/cc/etc., that sort of inconsistency is more common than you'd think/hope. And `serialize_data()` is the right place to fix it up. (Search for "late binding" or "late-bound" in the other backends to see [several examples](https://github.com/anymail/django-anymail/search?q=late&type=Code&utf8=%E2%9C%93) of this.)

You can certainly transform the existing data in serialize_data. Another approach I often use is to just [stash the raw values](https://github.com/anymail/django-anymail/blob/v1.4/anymail/backends/mailjet.py#L182-L186) in the payload (outside the data prop), and then [process them into data](https://github.com/anymail/django-anymail/blob/v1.4/anymail/backends/mailjet.py#L91-L107) in serialize_data, once I know what format is needed.

Once the backend tests are in place, I think we can merge this. The webhooks can come later, as a separate PR. (Also, I signed up for a SendinBlue account, so we'll be able to run your live integration tests in the Travis builds.)

medmunds commented on anymail/backends/sendinblue.py:
```diff
@@ -0,0 +1,200 @@
+        for key in self.data.keys():
+            if key in transformation.keys():
+                new_key = transformation[key]
+                self.data[new_key] = [email.get('email') for email in self.data.pop(key)]
```
I think this is going to lose the recipient's display_name when sending a template email.

What you probably want here is the fully-formatted [`email.address`](https://github.com/anymail/django-anymail/blob/v1.4/anymail/backends/mailjet.py#L91-L107) from the original EmailAddress objects passed to `set_recipients`.

RignonNoel replied:
SendinBlue doesn't support `display_name` on templated email :unamused: ...

https://developers.sendinblue.com/v3.0/reference#sendtemplate
```
curl --request POST \
  --url https://api.sendinblue.com/v3/smtp/templates/templateId/send \
  --data '{"emailTo":["to1@example.com","to2@example.com"]}'
```

RignonNoel commented:
Hi @medmunds, No problem for the delay we all have a life on the side! 

I just added some backend test for sendinblue. I hope it do the work for you, i tried to highlight all unsupported feature and limitation (no metadata, only one tag, ..)

I just have a very empty test for `template_id`. Have you a way to test that on this context ? For the moment it just test to send and check if no error are trigger but it's not the best check i've seen in my life :/

medmunds commented on tests/test_sendinblue_backend.py:
```diff
@@ -0,0 +1,444 @@
+        self.message.template_id = "12"
+        self.message.send()
+
+        # We can't verify if the Url has changed
```
`self.assert_esp_called('/v3/smtp/templates/12/send')` will check the URL. (See RequestsBackendMockAPITestCase.) 

This would also be a good place to verify that all the API parameters are transformed as expected for the template API call. ("to" --> "emailTo", etc.)

RignonNoel replied:
Done

medmunds commented on tests/test_sendinblue_backend.py:
```diff
@@ -0,0 +1,444 @@
+        self.assertNotIn('subject', data)
+
+    def test_merge_data(self):
+        self.message.from_email = 'from@example.com'
```
For most backends, this is a combined test for `message.merge_data` and `message.merge_global_data`. Since SendinBlue only supports the global variant, it'd be good to split that out (or roll it into the template test earlier), and verify that merge_global_data ends up as SiB "attributes". [Example from Postmark tests](https://github.com/anymail/django-anymail/blob/v1.4/tests/test_postmark_backend.py#L362-L377)

RignonNoel replied:
Done

medmunds commented on tests/test_sendinblue_backend.py:
```diff
@@ -0,0 +1,444 @@
+        self.assertNotIn('attachment', data)
+        self.assertNotIn('tag', data)
+        self.assertNotIn('headers', data)
+        self.assertNotIn('send_at', data)
```
This should assertNotIn for all the optional ESP API params. So, it's not necessary to check `'send_at'` (not a SiB param), but it'd be good to check `'replyTo'` and maybe `'attributes'`.

RignonNoel replied:
Done

medmunds commented on tests/test_sendinblue_backend.py:
```diff
@@ -0,0 +1,444 @@
+        sent = msg.send()
+        self.assertEqual(sent, 1)
+        self.assertEqual(msg.anymail_status.status, {'queued'})
+        self.assertEqual(msg.anymail_status.recipients['to1@example.com'].status, 'queued')
```
This should also check that the `msg.anymail_status.message_id` and `msg.anymail_status.recipients['to1@example.com'].message_id` get set to the message_id from SiB's API response.

RignonNoel replied:
Done

medmunds commented:
@RignonNoel backend tests look great—thanks! I'll add inline notes on testing template_id and a few other things.

I experimented a little more with the SendinBlue transactional APIs, and found a few things:
* You can set an X-Mailin-custom header to an arbitrary string, and it will be passed to webhooks. This would be a good way to implement Anymail's [metadata](https://anymail.readthedocs.io/en/stable/sending/anymail_additions/#anymail.message.AnymailMessage.metadata) extension, by passing it as JSON ([example](https://github.com/anymail/django-anymail/blob/v1.4/anymail/backends/mailjet.py#L224-L226)). The exact capitalization of the header name seems to matter, and SiB's docs aren't always consistent. Lowercase "custom" is the one that worked for me. (Ditto X-Mailin-tag, btw. X-Mailin-Tag with a capital T doesn't get passed to webhooks.)
* Attachment content-type is determined from the filename extension. (And trying to send an attachment without a name or a name without an extension generates a 400 response with an "Unsupported file format" message. I don't think the code needs to do anything special for this, but it's worth noting in the docs.)
* SendinBlue has a bunch of limitations that might not be expected coming from other ESPs (e.g., can't send inline images, can't specify display-names when sending a template, etc.). I'll add inline notes for the other ones I found. In all these cases, the payload needs to call `self.unsupported_feature("...")`. (Issuing errors on unsupported features is important for users trying to switch ESPs, in case they're counting on something that won't work.)

Let me know if you want to take a crack at those, otherwise I can fill them in.

Also, you're the first person other than me to fully implement a new backend for Anymail, so a huge thanks! I've learned a lot from this about places I should improve the internal documentation and porting notes, and I really appreciate the effort you put into powering through what's there now.

medmunds commented on anymail/backends/sendinblue.py:
```diff
@@ -0,0 +1,205 @@
+        message_id = None
+
+        if response.content != b'':
+            parsed_response = self.deserialize_json_response(response, payload, message)
```
Uh, sorry—ignore this comment. deserialize_json_response worries about that for you.

RignonNoel commented on tests/test_sendinblue_backend.py:
```diff
@@ -0,0 +1,444 @@
+
+    def test_embedded_images(self):
+        # SendinBlue doesn't support inline image
+        # inline image are just added as a content attachment
```
Done

RignonNoel commented:
@medmunds Thanks a lot for all your comments! It really help me to understand some part of anymail (i really look forward to use it now!! :smile: ).

 - I rebased all my commits to fix your inline comments
 - I added some details on SendinBlue documentation (I did it concisely because I still write pretty bad English...)
 - I added metadata support in the Payload and refactor attached test

Let me know if you have some other comments!

medmunds commented:
@RignonNoel outstanding, thanks! Merging now. I'll aim to release it in the next few days.

Let me know if you want to take a crack at SendinBlue webhooks.

medmunds commented:
Just a followup: SendinBlue support has confirmed that their "V3 [API does] not support inline images."
